Sources/PackageRegistry/RegistryManager.swift

```diff
@@ -0,0 +1,293 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import PackageLoading
+import PackageModel
+
+import TSCBasic
+import TSCUtility
+
+import struct Foundation.URL
+import struct Foundation.URLComponents
+import struct Foundation.URLQueryItem
+
+import Dispatch
+
+public enum RegistryError: Error {
+    case invalidOperation
+    case invalidResponse
+    case invalidURL
+    case invalidChecksum(expected: String, actual: String)
+}
+
+fileprivate let userAgent = "SwiftPackageManager/\(ToolsVersion.currentToolsVersion)"
+
+public final class RegistryManager {
+    internal static var clientFactory: () -> HTTPClientProtocol = {
+        var configuration = HTTPClientConfiguration()
+        configuration.followRedirects = false
+
+        return HTTPClient(configuration: configuration, handler: nil, diagnosticsEngine: nil)
+    }
+
+    let registryBaseURL: Foundation.URL
+
+    init(registryBaseURL: Foundation.URL) {
+        self.registryBaseURL = registryBaseURL
+    }
+
+    public class func discover(
+        for package: PackageReference,
+        completion: @escaping (Result<RegistryManager, Error>) -> Void
+    ) {
+        guard let url = URL(string: "https://\(package.identity)") else {
+            return completion(.failure(RegistryError.invalidURL))
+        }
+
+        let request = HTTPClient.Request(
+            method: .head,
+            url: url,
+            headers: [
+                "Accept": "application/vnd.swift.registry.v1",
+                "User-Agent": userAgent
+            ]
+        )
+
+        let client = clientFactory()
+        client.execute(request) { result in
+            switch result {
+            case .success(let response):
+                if response.statusCode == 303,
```

Comment:
the http client allows you to configure  "allowed response codes" which does this check for you

Reply:
I saw that, but I didn't see a way to conveniently configure that on a per-request basis as would be needed here.

Reply:
https://github.com/apple/swift-package-manager/blob/main/Tests/BasicsTests/HTTPClientTests.swift#L276